### PR TITLE
Factor out defining revisions into an async job

### PR DESCRIPTION
**Pull request by pavelkalinnikov**

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
First cut. PTAL.

**Comment by pavelkalinnikov** [comment]
Some integrations tests need tweaking.

File: core/sequencer/server.go [hunk 1]
```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
...
+	resp := &spb.DefineRevisionsResponse{
+		Begin: int64(root.Revision) + 1,
+		End:   high,
 	}
 
 	// Don't create new revisions if there are ones waiting to be applied.
-	if len(outstanding) > 0 {
-		return &spb.DefineRevisionsResponse{OutstandingRevisions: outstanding}, nil
+	// TODO(#1056): Allow going ahead to some extent.
+	if resp.End > resp.Begin {
```
Comment (pavelkalinnikov) [inline core/sequencer/server.go]: We probably should allow at lest 1 outstanding revision here, otherwise this PR will only make things 1.5 times slower.

File: core/sequencer/sequencer.go [hunk 2]
```diff
@@ -121,16 +121,31 @@ func (s *Sequencer) ForAllMasterships(ctx context.Context, f func(ctx context.Co
...
 }
 
-// RunBatchForAllMasterships runs KeyTransparencySequencerClient.RunBatch on all
-// directories this sequencer is currently master for.
-func (s *Sequencer) RunBatchForAllMasterships(ctx context.Context, batchSize int32) error {
-	glog.Infof("RunBatchForAllMasterships")
+// DefineRevisionsForAllMasterships runs KeyTransparencySequencerClient's
+// DefineRevisions method on all directories that this sequencer is currently
+// master for.
+func (s *Sequencer) DefineRevisionsForAllMasterships(ctx context.Context, batchSize int32) error {
```
Comment (gdbelvin) [inline core/sequencer/sequencer.go]: I think there's a more generic RunForAllMasterships function that might reduce some code duplication

File: core/sequencer/sequencer_api.proto [hunk 3]
```diff
@@ -75,8 +70,29 @@ message DefineRevisionsRequest {
...
+  // begin is the first outstanding revision (inclusive).
+  int64 begin = 2;
+  // end is the last outstanding revision (exclusive). It is equal to begin if
+  // there are no outstanding revisions.
+  int64 end = 3;
+}
+
+// GetNewRevisionsRequest contains information needed to query the set of
+// outstanding revisions.
+message GetNewRevisionsRequest {
```
Comment (jtoohill) [inline core/sequencer/sequencer_api.proto]: Maybe "GetUnappliedRevisionsRequest/Response" or "GetOutstandingRevisionsRequest/Response"? "New" could be a misnomer, as revisions may be defined but not applied for a long time. Whatever you pick, be consistent with the field naming above.

File: core/sequencer/sequencer_api.proto [hunk 4]
```diff
@@ -49,13 +49,8 @@ message MapMetadata {
...
   // directory_id is the directory to run for.
   string directory_id = 1;
-  // min_batch is the minimum number of items in a batch.
-  // If less than min_batch items are available, nothing happens.
-  // TODO(#1047): Replace with timeout so items in the log get processed
-  // eventually.
-  int32 min_batch = 2;
-  // max_batch is the maximum number of items in a batch.
-  int32 max_batch = 3;
+  // Reserved for the deprecated min_batch and max_batch fields.
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Done.

File: core/sequencer/sequencer_api.proto (hunk 3 above)
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Done.

File: core/sequencer/sequencer_api.proto [hunk 5]
```diff
@@ -75,8 +70,29 @@ message DefineRevisionsRequest {
 
 // DefineRevisionResponse contains information about freshly defined revisions.
 message DefineRevisionsResponse {
-  // outsanding_revisions a list of all the defined revisions which are not yet applied.
-  repeated int64 outstanding_revisions = 1;
+  // Reserved for the deprecated outstanding_revisions field.
+  reserved 1;
+  // begin is the first outstanding revision (inclusive).
+  int64 begin = 2;
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Done.

File: core/integration/client_tests.go [hunk 6]
```diff
@@ -72,11 +72,7 @@ func runSequencer(ctx context.Context, t *testing.T, dirID string, env *Env) {
 	ticker := time.NewTicker(100 * time.Millisecond)
 	defer ticker.Stop()
 	sequencer.PeriodicallyRun(ctx, ticker.C, func(ctx context.Context) {
```
Comment (gdbelvin) [inline core/integration/client_tests.go]: Do you also need to run `DefineRevision` in addition to `RunBatch` here?

**Comment by gdbelvin** [comment]
Real integration test is failing with `Error: update failed: context deadline exceeded` -- are all the right go routines getting launched?

File: core/sequencer/server.go [hunk 7]
```diff
@@ -304,20 +291,56 @@ func (s *Server) DefineRevisions(ctx context.Context,
...
+		Begin: int64(root.Revision) + 1,
+		End:   high,
+	}
+	if revs.End < revs.Begin { // In case there is a race.
+		revs.End = revs.Begin // Avoid ugly results.
+	}
+	return revs, nil
+}
+
+// getRootAndRev returns the latest map root, and the first unprocessed
```
Comment (gdbelvin) [inline core/sequencer/server.go]: The name is a bit confusing -- maybe because I made it do something confusing?
Perhaps this should be split into two functions?

LatestAppliedMapRevision
LatestDefinedBatchRevision

File: core/sequencer/sequencer_api.proto [hunk 8]
```diff
@@ -75,8 +70,29 @@ message DefineRevisionsRequest {
 
 // DefineRevisionResponse contains information about freshly defined revisions.
 message DefineRevisionsResponse {
-  // outsanding_revisions a list of all the defined revisions which are not yet applied.
-  repeated int64 outstanding_revisions = 1;
+  // unapplied_begin is the first outstanding revision (inclusive).
+  int64 unapplied_begin = 2;
```
Comment (gdbelvin) [inline core/sequencer/sequencer_api.proto]: I find this a bit harder to grok.  
Was there a particular reason for the change?

File: core/sequencer/server.go [hunk 9]
```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
...
+	// Get the list of unprocessed revisions.
+	root, high, err := s.getRootAndRev(ctx, in.DirectoryId)
 	if err != nil {
 		return nil, err
 	}
-	outstanding := []int64{}
-	for rev := int64(latestMapRoot.Revision) + 1; rev <= highestRev; rev++ {
-		outstanding = append(outstanding, rev)
+	resp := &spb.DefineRevisionsResponse{
+		UnappliedBegin: int64(root.Revision) + 1,
```
Comment (gdbelvin) [inline core/sequencer/server.go]:
```suggestion
		UnappliedBegin: int64(root.Revision) + 1,  // The next map revision to apply
```

File: core/sequencer/server.go [hunk 10]
```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
...
+		UnappliedEnd:   high,
```
Comment (gdbelvin) [inline core/sequencer/server.go]:
```suggestion
		UnappliedEnd:   high,  // The highest pre-defined revision from the batch definitions table. 
```

File: core/sequencer/server.go [hunk 11]
```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
...
-	}
-	_, latestMapRoot, err := mapClient.GetAndVerifyLatestMapRoot(ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	// Collect a list of unapplied revisions.
-	highestRev, err := s.batcher.HighestRev(ctx, in.DirectoryId)
+	// Get the list of unprocessed revisions.
+	root, high, err := s.getRootAndRev(ctx, in.DirectoryId)
```
Comment (gdbelvin) [inline core/sequencer/server.go]: Would it be better to call GetUnappliedRevisions here?

File: core/integration/client_tests.go [hunk 12]
```diff
@@ -523,8 +519,6 @@ func (env *Env) setupHistory(ctx context.Context, directory *pb.Directory, userI
 			}
 			if _, err := env.Sequencer.RunBatch(ctx, &spb.RunBatchRequest{
```
Comment (gdbelvin) [inline core/integration/client_tests.go]: And here and the other places in this file.  Maybe a helper function is needed

Comment (pavelkalinnikov) [reply core/integration/client_tests.go]: Done, a helper was added in #1341.

File: core/sequencer/sequencer_api.proto (hunk 8 above)
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: `outstanding_revisions` always contains a contiguous range of revisions, so this is it just to avoid waste of space. Let me see if I can detach this to a prereq PR.

File: core/sequencer/sequencer_api.proto [hunk 13]
```diff
@@ -75,8 +68,29 @@ message DefineRevisionsRequest {
...
+// outstanding revisions.
+message GetUnappliedRevisionsRequest {
+  // directory_id is the directory to examine the outstanding revisions for.
+  string directory_id = 1;
+}
+
+// GetUnappliedRevisionResponse contains the set of outstanding revisions.
+message GetUnappliedRevisionsResponse {
+  // begin is the first outstanding revision (inclusive).
+  int64 begin = 1;
```
Comment (gdbelvin) [inline core/sequencer/sequencer_api.proto]: Nit: these names should match the names in DefineRevisionsResponse

File: core/sequencer/server.go [hunk 14]
```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
...
-		MinBatch:    in.MinBatch,
-		MaxBatch:    in.MaxBatch})
+	revs, err := s.loopback.GetUnappliedRevisions(ctx,
+		&spb.GetUnappliedRevisionsRequest{DirectoryId: in.DirectoryId})
 	if err != nil {
 		return nil, err
 	}
 
-	unappliedRevisions.Set(float64(len(defResp.OutstandingRevisions)))
+	count, end := revs.End-revs.Begin, revs.End
```
Comment (gdbelvin) [inline core/sequencer/server.go]: This changes the behavior slightly -- could this be it's own PR?

File: core/sequencer/server.go (hunk 7 above)
Comment (pavelkalinnikov) [reply core/sequencer/server.go]: Not relevant after another refactoring.

File: core/sequencer/server.go (hunk 9 above)
Comment (pavelkalinnikov) [reply core/sequencer/server.go]: Not relevant after another refactoring.

File: core/sequencer/server.go (hunk 14 above)
Comment (pavelkalinnikov) [reply core/sequencer/server.go]: I think the metric should be more or less the same.

File: core/sequencer/sequencer_api.proto (hunk 13 above)
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Got back to lists of revisions.

File: core/sequencer/sequencer_api.proto (hunk 8 above)
Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Using revision lists again.

File: cmd/keytransparency-sequencer/main.go [hunk 15]
```diff
@@ -231,11 +231,17 @@ func runSequencer(ctx context.Context, conn *grpc.ClientConn,
 	if err := signer.AddAllDirectories(ctx); err != nil {
 		glog.Errorf("runSequencer(AddAllDirectories): %v", err)
 	}
+
 	go sequencer.PeriodicallyRun(ctx, time.Tick(*refresh), func(ctx context.Context) {
 		if err := signer.AddAllDirectories(ctx); err != nil {
 			glog.Errorf("PeriodicallyRun(AddAllDirectories): %v", err)
 		}
-		if err := signer.RunBatchForAllMasterships(ctx, int32(*batchSize)); err != nil {
+		if err := signer.DefineRevisionsForAllMasterships(ctx, int32(*batchSize)); err != nil {
```
Comment (pavelkalinnikov) [inline cmd/keytransparency-sequencer/main.go]: Should we separate this from `AddAllDirectories`, to save a bit of latency?

Comment (pavelkalinnikov) [reply cmd/keytransparency-sequencer/main.go]: I think `AddAllDirectories` should also run with a more relaxed cadence, because directories are rarely (almost never) added.

Comment (pavelkalinnikov) [reply cmd/keytransparency-sequencer/main.go]: #1348

File: core/sequencer/sequencer_api.proto [hunk 16]
```diff
@@ -75,8 +68,23 @@ message DefineRevisionsRequest {
...
-  repeated int64 outstanding_revisions = 1;
+  // revisions is the list of newly defined revisions.
+  repeated int64 revisions = 2;
+  // Deprecated tags, do not reuse.
+  reserved 1;
+}
+
+// GetUnappliedRevisionsRequest contains information needed to query the set of
+// outstanding revisions.
+message GetUnappliedRevisionsRequest {
```
Comment (gdbelvin) [inline core/sequencer/sequencer_api.proto]: `ListUnappliedRevisions`

Comment (gdbelvin) [reply core/sequencer/sequencer_api.proto]: PageSize

Comment (pavelkalinnikov) [reply core/sequencer/sequencer_api.proto]: Not relevant anymore, as we decided to remove lists of revisions.

**Comment by pavelkalinnikov** [comment]
Since this PR has gone through many transformations/merges, I will squash it and rebase on #1349.

**Comment by pavelkalinnikov** [comment]
Please take another look.
